experimental/ActionPolyRing/src/Content.jl

```diff
@@ -0,0 +1,1373 @@
+    coeff_t = coeff(term, 1)
+    vars_t = vars(term)
+    ev = append!(exponent_vector(term, 1), fill(0, ngens(upr) - length(exponent_vector(term, 1))))
+    new_exp_vec = copy(ev)
+
+    for var in vars_t
+      var_pos = findfirst(==(var), gens(upr))
+      shifted_var_pos = old_to_new_pos[var_pos]
+      
+      new_exp_vec[shifted_var_pos] = ev[var_pos]
+      new_exp_vec[var_pos] = 0 
+    end
+    
+    push_term!(C, coeff_t , new_exp_vec)
+  end
+  return dpr(finish(C))
+end
+
+function diff_action(dpre::DifferentialPolyRingElem{T}, d::Vector{Int}) where {T}
+  len = length(d)
+  @req len == ndiffs(parent(dpre)) && all(j -> j >= 0, d) "Invalid vector of diff multiplicities"
+  res = dpre
+  for i in 1:len
+    for j in 1:d[i]
+      res = diff_action(res, i)
+    end
+  end
+  return res
+end
+
+@doc"""
+    initial(p::ActionPolyRingElem)
+
+Return the initial of the polynomial `p`, i.e. the leading coefficient of `p` regarded as a univariate polynomial in its leader.
+"""
+function initial(apre::ActionPolyRingElem)
+  if is_constant(apre)
+    return apre
+  end
+  return divexact(leading_term(apre), leader(apre)^degree(apre, leader(apre)); check = true)
+end
+
+@doc raw"""
+    leader(p::ActionPolyRingElem)
+
+Return the leader of the polynomial `p`, that is the largest variable with respect to the ranking of `parent(p)`. If `p` is constant, an error is raised.
+"""
+function leader(apre::ActionPolyRingElem)
+  @req !is_constant(apre) "A constant polynomial has no leader"
+  if is_univariate(apre)
+    return vars(apre)[1]
+  end
+  return max(vars(apre)...)
+end
+###############################################################################
+#
+#   Iterators
+#
+###############################################################################
+
+function coefficients(a::DifferencePolyRingElem)
+   return DifferencePolyCoeffs(a)
+end
+
+function coefficients(a::DifferentialPolyRingElem)
+   return DifferentialPolyCoeffs(a)
+end
+
+function AbstractAlgebra.exponent_vectors(a::DifferencePolyRingElem)
+   return DifferencePolyExponentVectors(a)
+end
+
+function AbstractAlgebra.exponent_vectors(a::DifferentialPolyRingElem)
+   return DifferentialPolyExponentVectors(a)
+end
+
+function monomials(a::DifferencePolyRingElem)
+   return DifferencePolyMonomials(a)
+end
+
+function monomials(a::DifferentialPolyRingElem)
+   return DifferentialPolyMonomials(a)
+end
+
+function terms(a::DifferencePolyRingElem)
+   return DifferencePolyTerms(a)
+end
+
+function terms(a::DifferentialPolyRingElem)
+   return DifferentialPolyTerms(a)
+end
+
+function Base.iterate(x::Union{DifferencePolyCoeffs, DifferentialPolyCoeffs})
+   if length(x.poly) >= 1
+      return coeff(x.poly, 1), 1
+   else
+      return nothing
+   end
+end
+
+function Base.iterate(x::Union{DifferencePolyCoeffs, DifferentialPolyCoeffs}, state)
+   state += 1
+   if length(x.poly) >= state
+      return coeff(x.poly, state), state
+   else
+      return nothing
+   end
+end
```

Comment:
```suggestion
function Base.iterate(x::Union{DifferencePolyCoeffs, DifferentialPolyCoeffs}, state=0)
   state += 1
   if length(x.poly) >= state
      return coeff(x.poly, state), state
   else
      return nothing
   end
end
```
reduce code duplication. the exact same is possible for all of these `iterate` methods here